README.md
# **e-keyboard**

## Qt virtual Keyboard

It is an advanced virtual keyboard based on x11 keycode relations and server-client programming for UNIX like systems. Major Features are:

* Floating
* Fast
* Ability to change system wide keyboard layout
* dbus interface can be reached in session
* full and tablet keyboard layout support
* EOE (easy on eyes) color themes
* Easy to configure with the help of self contain settings


### Build

Clone the project
```bash
git clone https://github.com/bayramkarahan/e-keyboard.git
```
Install build dependencies
```bash
sudo apt install build-essential libc6 libgcc1 libgl1-mesa-glx | libgl1 libqt5core5a libqt5dbus5 libqt5gui5  libqt5network5 libqt5qml5 libqt5quick5 libqt5svg5-dev libqt5widgets5 libqt5x11extras5-dev libstdc++6 libx11-6 libx11-xcb-dev libxcb-xkb-dev libxcb1-dev libxkbcommon-x11-0 libxkbcommon-x11-dev libxkbcommon0   libxkbcommon-dev libxkbfile-dev libxtst-dev qtdeclarative5-dev
```

Build
```bash
cd e-keyboard
mkdir build
cd build
qmake ../
make
```

Install Runtime dependencies
```bash
sudo apt install libqt5svg5 qml-module-qtquick-controls qml-module-qtquick-window2 qml-module-qtquick2
```

Run
```bash
./e-keyboard show
```

You can also test remote calling
```bash
sudo apt install qdbus-qt5
qdbus org.eta.virtualkeyboard /VirtualKeyboard org.eta.virtualkeyboard.toggle
```

### Debian packaging

```bash
cd e-keyboard
sudo apt install build-essential debmake debhelper git-buildpackage
sudo mk-build-deps -ir
gbp buildpackage --git-export-dir=/tmp/build-area -b -us -uc
cd /tmp/build-area
```


debian/changelog
e-keyboard (1.0.0) unstable; urgency=medium

  * Initial Release

 -- Bayram KARAHAN <bayramk@gmail.com>  Sun, 27 Dec 2020 14:31:08 +0300


src/logger.cpp
#include "logger.h"
#include <QDir>
#include <QDateTime>
#include <QtWidgets>
#include <QFile>
#include <QFileInfo>
#include <QTextStream>
#include <QCoreApplication>

#define log_file_path "/var/log"
#define log_file_name "e.log"

Logger::Logger(QObject *parent) : QObject(parent)
{
    m_path = QCoreApplication::applicationFilePath();

    QStringList homePath =
            QStandardPaths::standardLocations(QStandardPaths::HomeLocation);
    whoami = homePath.first().split(QDir::separator()).last();

}

QString Logger::getPath() const
{
    return m_path;
}

QString Logger::getUser() const
{
    return whoami;
}

void Logger::log(const QString &str)
{
    createFile(log_file_path, log_file_name);
    file.setFileName(file_fullpath);
    if (file.open(QIODevice::WriteOnly | QIODevice::Append)){
        if (writeData(str)) {
            file.close();
        }
        else qDebug() << "Log information could not write to log file";
    } else {
        qDebug() << "Log information could not open";
    }
}

bool Logger::createFile(const QString &file_path, const QString &file_name)
{
    file_fullpath = file_path + "/" + file_name;
    QFileInfo checkFile(file_fullpath);
    d = new QDir(file_path);
    if(!checkFile.exists() || !checkFile.isFile()) {
        d->mkpath(file_path);
        QFile file(file_fullpath);
        if (file.open(QIODevice::ReadWrite)) {
            file.setPermissions(QFileDevice::ReadOther |
                                QFileDevice::WriteOther |
                                QFileDevice::ReadGroup |
                                QFileDevice::WriteGroup |
                                QFileDevice::ReadOwner |
                                QFileDevice::WriteOwner |
                                QFileDevice::ExeOwner);
            file.close();
            return true;
        } else {
            qDebug() << file.errorString();
        }
    }
    return false;

}

void Logger::updateTime()
{
    QDateTime now = QDateTime::currentDateTime();
    m_current_time = now.toString("dd-MM-yyyy hh:mm:ss.zzz");
}

bool Logger::writeData(const QString &str)
{
    try {
        updateTime();
        QTextStream out(&file);
        out << m_current_time + " " + yellow_color + getPath() + no_color + " : " + str + "\n";
        return true;
    } catch (...) {
        return false;
    }
}


src/settings.cpp
/*****************************************************************************
 *   Copyright (C) 2016 by Yunusemre Senturk                                 *
 *   <yunusemre.senturk@pardus.org.tr>                                       *
 *                                                                           *
 *   This program is free software; you can redistribute it and/or modify    *
 *   it under the terms of the GNU General Public License as published by    *
 *   the Free Software Foundation; either version 2 of the License, or       *
 *   (at your option) any later version.                                     *
 *                                                                           *
 *   This program is distributed in the hope that it will be useful,         *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           *
 *   GNU General Public License for more details.                            *
 *                                                                           *
 *   You should have received a copy of the GNU General Public License       *
 *   along with this program; if not, write to the                           *
 *   Free Software Foundation, Inc.,                                         *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .          *
 *****************************************************************************/
#include "src/settings.h"
#include <QDir>
#include <QFileInfo>
#include <QSettings>
#include <QVariant>
#include <QString>

Settings::Settings(QObject *parent) :
    QObject(parent),
    m_color(0),
    m_scale(1.0),
    m_autoShow(true),
    m_opacity(1.0)
{
    configpath = QDir::homePath() + "/.config/virtualkeyboard/config.ini";

    QFileInfo checkConfig(configpath);

    if (checkConfig.exists() && checkConfig.isFile()) {
        preferences = new QSettings(configpath,QSettings::IniFormat);
        preferences->beginGroup("virtualkeyboard");
        m_color = preferences->value("Color").toInt();
        m_layoutType = preferences->value("LayoutType").toString();
        m_scale = preferences->value("Scale").toDouble();
        m_languageLayoutIndex = preferences->
                value("LanguageLayoutIndex").toUInt();
        m_autoShow = preferences->value("AutoShow").toBool();
        m_opacity = preferences->value("Opacity").toDouble();
        preferences->endGroup();
    }
}

void Settings::setSettings(const int color, const QString& layoutType,
                           double scale,
                           unsigned int languageLayoutIndex, bool autoShow,
                           double opacity)
{
    this->m_color = color;
    this->m_layoutType = layoutType;
    this->m_scale = scale;
    this->m_languageLayoutIndex = languageLayoutIndex;
    this->m_autoShow = autoShow;
    this->m_opacity = opacity;
}

int Settings::getColor() const
{
    return this->m_color;
}

QString Settings::getLayoutType() const
{
    return this->m_layoutType;
}

double Settings::getScale()
{
    return this->m_scale;
}

unsigned int Settings::getLanguageLayoutIndex()
{
    return this->m_languageLayoutIndex;
}

bool Settings::getAutoShow()
{
    return this->m_autoShow;
}

double Settings::getOpacity()
{
    return this->m_opacity;
}

void Settings::saveSettings()
{
    preferences = new QSettings(configpath,QSettings::IniFormat);
    preferences->beginGroup("virtualkeyboard");
    QVariant varColor(this->m_color);
    preferences->setValue("Color",varColor);
    QVariant varLayoutType(this->m_layoutType);
    preferences->setValue("LayoutType",varLayoutType);
    QVariant varScale(this->m_scale);
    preferences->setValue("Scale",varScale);
    QVariant varLanguage(this->m_languageLayoutIndex);
    preferences->setValue("LanguageLayoutIndex",varLanguage);
    QVariant varAutoShow(this->m_autoShow);
    preferences->setValue("AutoShow",varAutoShow);
    QVariant varOpacity(this->m_opacity);
    preferences->setValue("Opacity",varOpacity);
    preferences->endGroup();
}


src/vkdbusinterface.cpp
/*****************************************************************************
 *   Copyright (C) 2016 by Yunusemre Senturk                                 *
 *   <yunusemre.senturk@pardus.org.tr>                                       *
 *                                                                           *
 *   This program is free software; you can redistribute it and/or modify    *
 *   it under the terms of the GNU General Public License as published by    *
 *   the Free Software Foundation; either version 2 of the License, or       *
 *   (at your option) any later version.                                     *
 *                                                                           *
 *   This program is distributed in the hope that it will be useful,         *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           *
 *   GNU General Public License for more details.                            *
 *                                                                           *
 *   You should have received a copy of the GNU General Public License       *
 *   along with this program; if not, write to the                           *
 *   Free Software Foundation, Inc.,                                         *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .          *
 *****************************************************************************/
#include "src/vkdbusinterface.h"
#include "src/adaptor.h"
#include <QDBusConnection>

VkDbusInterface::VkDbusInterface(QObject *parent) :
    QObject(parent)
{
    new VirtualKeyboardInterfaceAdaptor(this);
    QDBusConnection connection = QDBusConnection::sessionBus();
    connection.registerObject("/VirtualKeyboard", this);
    connection.registerService("org.e.virtualkeyboard");
}

void VkDbusInterface::showSlot(bool password)
{
    emit show(password);
}

void VkDbusInterface::toggleSlot()
{
    emit toggle();
}
void VkDbusInterface::hideSlot()
{
    emit hide();
}
void VkDbusInterface::fromLeftSlot()
{
    emit showFromLeft();
}
void VkDbusInterface::fromRightSlot()
{
    emit showFromRight();
}
void VkDbusInterface::fromBottomSlot()
{
    emit showFromBottom();
}
void VkDbusInterface::toggleAutoShowSlot()
{
    emit toggleAutoShow();
}
void VkDbusInterface::showPinInputSlot()
{
    emit showPinInput();
}
void VkDbusInterface::hidePinInputSlot()
{
    emit hidePinInput();
}


src/main.cpp
/*****************************************************************************
 *   Copyright (C) 2016 by Yunusemre Senturk                                 *
 *   <yunusemre.senturk@pardus.org.tr>                                       *
 *                                                                           *
 *   This program is free software; you can redistribute it and/or modify    *
 *   it under the terms of the GNU General Public License as published by    *
 *   the Free Software Foundation; either version 2 of the License, or       *
 *   (at your option) any later version.                                     *
 *                                                                           *
 *   This program is distributed in the hope that it will be useful,         *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of          *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           *
 *   GNU General Public License for more details.                            *
 *                                                                           *
 *   You should have received a copy of the GNU General Public License       *
 *   along with this program; if not, write to the                           *
 *   Free Software Foundation, Inc.,                                         *
 *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA .          *
 *****************************************************************************/
#include "src/helper.h"
#include "src/singleinstance.h"
#include <signal.h>
#include <unistd.h>
#include <QApplication>
#include <QQmlApplicationEngine>
#include <QtQml>
#include <QDir>
#include <QCursor>

#define SINGLE_INSTANCE ".virtualkeyboard"
static int setup_unix_signal_handlers();

int main(int argc, char *argv[])
{
    if (argc == 2 && QString(argv[1]) == "login") {
        Helper::login = true;
    } else if (argc == 2 && QString(argv[1]) == "show") {
        Helper::showOnStart = true;
    }

    qmlRegisterType<Helper>("e.helper",1,0,"Helper");
    QApplication app(argc, argv);

    ///app.setOverrideCursor(QCursor(Qt::BlankCursor));

    QString pidName = SINGLE_INSTANCE;
    QString username = qgetenv("USER");
    if (username.isEmpty())
        username = qgetenv("USERNAME");
    QString tmpPath= "/tmp/";
    QString pidPath = tmpPath.append(username);


    QDir dir(pidPath);
    if (!dir.exists()) {
        dir.mkpath(".");
    }

    QString name = pidPath.append("/").append(pidName);

    SingleInstance cInstance;    

    if(cInstance.hasPrevious(name, QCoreApplication::arguments()))
    {
        if (argc == 2 && QString(argv[1]) == "show") {
            qDebug("Trying to show");
            return system("qdbus org.e.virtualkeyboard /VirtualKeyboard "
                          "org.e.virtualkeyboard.showFromBottom");
        } else {
            qDebug("eta-keyboard is allready running");
            return 0;
        }

    }

    if (cInstance.listen(name)) {
        qDebug() << "Creating single instance";
        setup_unix_signal_handlers();
    } else {
        qFatal("Couldn't create single instance aborting");
    }

    QQmlApplicationEngine engine;
    engine.load(QUrl(QStringLiteral("qrc:/ui/main.qml")));

    return app.exec();
}

static void handle_signal(int sig)
{
    QString pidName = SINGLE_INSTANCE;
    QString username = qgetenv("USER");
    if (username.isEmpty())
        username = qgetenv("USERNAME");
    QString tmpPath= "/tmp/";
    QString pidPath = tmpPath.append(username);
    QString name = pidPath.append("/").append(pidName);
    QByteArray ba = name.toLatin1();
    Q_UNUSED(sig);
    unlink(ba.data());
    exit(0);
}

static int setup_unix_signal_handlers()
{
    struct sigaction sig;    
    sig.sa_handler = handle_signal;
    sigemptyset(&sig.sa_mask);
    sig.sa_flags = 0;
    sig.sa_flags |= SA_RESTART;
    if (sigaction(SIGINT, &sig, 0)) {
        return 1;
    }
    if (sigaction(SIGTERM, &sig, 0)) {
        return 2;
    }
    return 0;
}
